frontends/kernel-seeder/seeder_main.c

```diff
@@ -0,0 +1,247 @@
+			    rpi->entropy_count);
+	}
+
+	close(esdm_rpcs_linux_fd);
+
+	return -errsv;
+}
+
+static void usage(void)
+{
+	esdm_logger(
+		LOGGER_WARN, LOGGER_C_SEEDER,
+		"esdm-kernel-seeder [--interval SECS] [--help] [-verbosity]\n");
+}
+
+/*
+ * Helper tool to insert entropy into the kernel RNG occasionally. When the IRQ ES
+ * is present, this is required as the kernel RNG is deprived of its main ES.
+ * But also in any other case it is good to insert data into the kernel RNG
+ * to provide data that is gathered from other entropy sources. Basically
+ * the ESDM acts as an RNGd to top up the entropy in the kernel.
+ */
+int main(int argc, char **argv)
+{
+	int c = 0;
+	int64_t seeding_interval_secs = ESDM_LINUX_RESEED_INTERVAL_SEC;
+	bool help = false;
+	int tool_ret = EXIT_SUCCESS;
+	int verbosity = 0;
+
+	while (1) {
+		int opt_index = 0;
+		static struct option opts[] = { { "interval", 1, 0, 0 },
+						{ "help", 0, 0, 0 },
+						{ "verbosity", 0, 0, 0 },
+						{ "syslog", 0, 0, 0 },
+						{ 0, 0, 0, 0 } };
+		c = getopt_long(argc, argv, "i:hvs", opts, &opt_index);
+		if (-1 == c)
+			break;
+		switch (c) {
+		case 0:
+			switch (opt_index) {
+			case 0:
+				/* seeding interval */
+				seeding_interval_secs =
+					strtoll(optarg, NULL, 10);
+				break;
+			case 1:
+				/* help */
+				help = true;
+				break;
+			case 2:
+				/* verbosity */
+				verbosity++;
+				break;
+			case 3:
+				/* syslog */
+				esdm_logger_enable_syslog("esdm-kernel-seeder");
+				break;
+			}
+			break;
+		case 'i':
+			seeding_interval_secs = strtoll(optarg, NULL, 10);
+			break;
+		case 'h':
+			help = true;
+			break;
+		case 'v':
+			verbosity++;
+			break;
+		case 's':
+			esdm_logger_enable_syslog("esdm-kernel-seeder");
+			break;
+		}
+	}
+
+	if (geteuid()) {
+		esdm_logger_inc_verbosity();
+		esdm_logger(LOGGER_ERR, LOGGER_C_SEEDER,
+			    "Program must start as root!\n");
+		return 1;
+	}
+
+	esdm_logger_set_verbosity(verbosity);
+
+	if (help) {
+		usage();
+		tool_ret = EXIT_FAILURE;
+		goto out;
+	}
+
+	uint8_t rpi_buf[sizeof(struct rand_pool_info) +
+			ESDM_SERVER_LINUX_ENTROPY_BYTES];
+	struct rand_pool_info *rpi = (struct rand_pool_info *)rpi_buf;
+
+	/* Wake up every 2 minutes by default */
+	struct timespec ts = { .tv_sec = seeding_interval_secs, .tv_nsec = 0 };
+	ssize_t ret;
+
+	rpi->buf_size = ESDM_SERVER_LINUX_ENTROPY_BYTES;
+
+	if (esdm_rpcc_init_unpriv_service(NULL) != 0) {
+		printf("unable to initialize unprivileged ESDM service, exiting!");
+		tool_ret = EXIT_FAILURE;
+		goto out;
+	}
+
+	for (;;) {
+		ret = esdm_rpcc_get_random_bytes_full((uint8_t *)rpi->buf,
+						      (size_t)rpi->buf_size);
+		if (ret < 0) {
+			esdm_logger(LOGGER_ERR, LOGGER_C_SEEDER,
+				    "Failure in generating random bits: %zd\n",
+				    ret);
+		} else {
+			rpi->entropy_count = ESDM_LINUX_RESEED_ENTROPY_COUNT;
+			esdm_rpcs_linux_insert_entropy(rpi);
+		}
+
+		explicit_bzero(rpi->buf, (size_t)rpi->buf_size);
```

Comment:
What is the reason to not use the ESDM-internal memset_secure?

Reply:
Corrected to memset_secure. No real reason. Was just too stupid to include the correct header I guess.